onnx/defs/shape_inference.h

```diff
@@ -224,6 +224,8 @@ void propagateElemTypeWithValidation(const TypeProto* input_type, TypeProto* out
 
 void propagateElemTypeFromInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex);
 
+void propagateElemTypeFromTensorInputToOutput(InferenceContext& ctx, size_t inputIndex, size_t outputIndex);
+
```

Comment:
For others' reference: This function was moved to .cc by my PR: https://github.com/onnx/onnx/pull/3772. However, ORT training relies on it: https://github.com/microsoft/onnxruntime/blob/6a7d3deb22bd78c2c243e1c8143000663127be5d/orttraining/orttraining/core/graph/training_op_defs.cc#L869.